### PR TITLE
Charter proposal for app-development-wg 

### PR DESCRIPTION
**Pull request by salaboy**
This PR includes the content of the Charter definition for the Application Development Working Group. The content was discussed in the following google doc: https://docs.google.com/document/d/1N_47QytSqi8-PW4wjnLBN5EDEArC_YDT5QgfN-r7nSY/edit 

In parallel, there currently a nomination for the Working Group Chairs here: https://github.com/cncf/tag-app-delivery/issues/627 

@ThomasVitale @danieloh30 please check if everything looks ok to you folks. .

@lianmakesthings @thschue Also, let me know if the structure is right or if changes are needed.

### CODE REVIEW

**Comment by danieloh30** [comment]
@salaboy Shouldn't the chapter document include only the content based on the standard template rather than internal discussions (e.g., Deliverables, Target Personas)? I think we need to remove the pages 9 - 14.

**Comment by salaboy** [comment]
@danieloh30 that is correct.. check the PR content: https://github.com/cncf/tag-app-delivery/pull/638/files

**Comment by danieloh30** [comment]
> @danieloh30 that is correct.. check the PR content: https://github.com/cncf/tag-app-delivery/pull/638/files

Cool! I didn't see the content PR in the PR :)

File: app-development-wg/charter/README.md [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+
+### Deliverables to the TAG
+
+This working group aims to bring developers closer to the App Delivery TAG. Creating resources and collaborating on initiatives that matter to cloud-native developers and architects. Concrete initiatives and deliverables that this working group will concentrate on: 
+
+Complementing the CNCF landscape with a view targeted for cloud-native developers and architects
+Complementing the Platform WG Glossary for Application Development
+Complementing other working groups under the TAG App Delivery personas definitions
+Helping with the organization and selection of proposals for AppDeveloperCon
```
Comment (ThomasVitale) [inline app-development-wg/charter/README.md:110]: Missing the bullets to render this as a list (see: https://github.com/cncf/tag-app-delivery/blob/b69babfe7897c50c6c2cfc3be88855e4965a1335/app-development-wg/charter/README.md)

```suggestion
* Complementing the CNCF landscape with a view targeted for cloud-native developers and architects
* Complementing the Platform WG Glossary for Application Development
* Complementing other working groups under the TAG App Delivery personas definitions
* Helping with the organization and selection of proposals for AppDeveloperCon
```

**Comment by ThomasVitale** [comment]
@salaboy it looks good! I only left a comment about formatting.

**Comment by salaboy** [comment]
@ThomasVitale thanks a lot for that! I was waiting to check the rendering

**Comment by salaboy** [comment]
@danieloh30 @ThomasVitale I tried to make the diagrams a bit better.. I think that they are a bit clearer now.. but there is a lot of room for improvements

File: app-development-wg/README.md
Comment (lianmakesthings) [inline app-development-wg/README.md:1]: I would suggest to add basic but important information here, i.e.:
- short (1 -2 sentence) description of the scope
- meeting time
- chairs

Comment (salaboy) [reply app-development-wg/README.md:1]: done!

File: app-development-wg/charter/README.md [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+- Ephemeral (remote) cloud-native runtime: a virtual scaled-down environment can be provisioned on-demand for a team to work on their development tasks. These environments are easy to create and easy to dispose of when they are no longer needed.
+  - If Kubernetes is being used, a platform team will be in charge of making the process simple and cost-efficient. 
+- Full remote cloud-native runtime: provide developers access to remote cloud-native runtime, where all the tools needed for their tasks are hosted remotely.
+  - Similar to cloud provider tooling, tools, including Kubernetes clusters, build tools, and code editors (IDEs), can all be hosted remotely, removing the need for developers to install tools locally. 
+
+![](imgs/environments.png)
+
+Depending on the context, these options can improve the development experience but still expose some inherent complexity of the cloud-native runtime environment of choice (for example, Kubernetes).
```
Comment (TheFoxAtWork) [inline app-development-wg/charter/README.md:28]:
```suggestion
Depending on the context, these options can improve the development experience but still expose some inherent complexity of the cloud native runtime environment of choice (for example, Kubernetes). 
```

File: app-development-wg/charter/README.md [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+- Researching different tools and approaches is a full-time job usually performed by other teams (such as platform teams)
+- How to enable developers across languages with out-of-the-box capabilities that they can use in their applications.
+
+
+While the rise of Kubernetes and containers has equipped teams with container orchestration expertise, a significant gap remains:  industry-wide standards for promoting applications across diverse environments. Standardization is crucial for protecting investments and justifying tooling choices.  For instance, the consistency of Kubernetes APIs across cloud providers allows teams to leverage their container orchestration skills regardless of the chosen cloud provider.  Similarly, the OpenTelemetry specification and tools provide a vendor-neutral approach to defining observability stacks based on the selected environment. However, the industry lacks a unified approach regarding application-level APIs, SDKs, and tools that can be used from different languages. This lack of shared best practices, APIs, and standards creates challenges and generates a lot of work that lands on cloud-native developers and architects' shoulders.
```
Comment (TheFoxAtWork) [inline app-development-wg/charter/README.md:37]:
```suggestion
While the rise of Kubernetes and containers has equipped teams with container orchestration expertise, a significant gap remains:  industry-wide standards for promoting applications across diverse environments. Standardization is crucial for protecting investments and justifying tooling choices.  For instance, the consistency of Kubernetes APIs across cloud providers allows teams to leverage their container orchestration skills regardless of the chosen cloud provider.  Similarly, the OpenTelemetry specification and tools provide a vendor-neutral approach to defining observability stacks based on the selected environment. However, the industry lacks a unified approach regarding application-level APIs, SDKs, and tools that can be used from different languages. This lack of shared best practices, APIs, and standards creates challenges and generates a lot of work that lands on cloud native developers and architects' shoulders.
```

File: app-development-wg/charter/README.md [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+
+## Mission
+
+The following sections describe the focus of the working group, its core responsabilities and the areas that are out of scope. 
+
+## Responsibilities & Deliverables
+
+Cloud-native developers and architects face a double threat: fragmented tooling landscapes and limited awareness of cross-language solutions. This combination creates a time-consuming roadblock – wasting valuable effort to find the right tools for common programming challenges.
```
Comment (TheFoxAtWork) [inline app-development-wg/charter/README.md:46]:
```suggestion
Cloud native developers and architects face a double threat: fragmented tooling landscapes and limited awareness of cross-language solutions. This combination creates a time-consuming roadblock – wasting valuable effort to find the right tools for common programming challenges.
```

File: app-development-wg/charter/README.md [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+This working group aims to empower cloud-native developers and architects by identifying and promoting common practices and tooling for enhancing the inner development loop. One of the objectives will be to provide an overview of categorized tools. Each tool, best practice, and standard the working group covers will highlight real-world case studies, suggestions, and content to help teams move faster. The key categories are below:
```
Comment (TheFoxAtWork) [inline app-development-wg/charter/README.md:48]:
```suggestion
This working group aims to empower cloud native developers and architects by identifying and promoting common practices and tooling for enhancing the inner development loop. One of the objectives will be to provide an overview of categorized tools. Each tool, best practice, and standard the working group covers will highlight real-world case studies, suggestions, and content to help teams move faster. The key categories are below:
```

File: app-development-wg/charter/README.md [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+
+The following list of goals will drive the working group agenda:
+
+- Identify and promote developer productivity tools within the cloud-native ecosystem. Raise awareness and facilitate the adoption of tools supporting and enhancing developer workflows in the cloud-native space.
+- Promote interoperability and standards across tools. Leverage shared APIs and SDKs across the industry to map the development ecosystem, document existing functionalities, and pinpoint areas for improvement.
+- Advocate for developer experience (DevEx) tools that streamline cloud-native development workflows.
+Assist with the organization of AppDeveloperCon, reviewing and proposing topics.
+- Identify how tools contribute to improving developer productivity metrics.
+- Provide early feedback and recommendations for projects and tools seeking CNCF adoption around cloud-native app development.
```
Comment (TheFoxAtWork) [inline app-development-wg/charter/README.md:88]:
```suggestion
- Identify and promote developer productivity tools within the cloud native ecosystem. Raise awareness and facilitate the adoption of tools supporting and enhancing developer workflows in the cloud native space.
- Promote interoperability and standards across tools. Leverage shared APIs and SDKs across the industry to map the development ecosystem, document existing functionalities, and pinpoint areas for improvement.
- Advocate for developer experience (DevEx) tools that streamline cloud native development workflows.
Assist with the organization of AppDeveloperCon, reviewing and proposing topics.
- Identify how tools contribute to improving developer productivity metrics.
- Provide early feedback and recommendations for projects and tools seeking CNCF adoption around cloud native app development. 
```

File: app-development-wg/charter/README.md [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+- Promote tools that are only useful for a specific stack/programming language.
+- Promote adopting security best practices throughout the development lifecycle, secure coding standards, and vulnerability scanning to mitigate risks and protect applications from potential threats.
+- Promote sustainability and efficiency in application development and tooling to optimize resource utilization and minimize environmental impact.
+
+
+### Deliverables to the TAG
+
+This working group aims to bring developers closer to the App Delivery TAG. Creating resources and collaborating on initiatives that matter to cloud-native developers and architects. Concrete initiatives and deliverables that this working group will concentrate on: 
+
+* Complementing the CNCF landscape with a view targeted for cloud-native developers and architects
```
Comment (TheFoxAtWork) [inline app-development-wg/charter/README.md:107]:
```suggestion
This working group aims to bring developers closer to the App Delivery TAG. Creating resources and collaborating on initiatives that matter to cloud native developers and architects. Concrete initiatives and deliverables that this working group will concentrate on: 

* Complementing the CNCF landscape with a view targeted for cloud native developers and architects
```

File: app-development-wg/charter/README.md [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+
+The following diagram illustrates common tools, tasks, and artifacts in developer workflows.
+
+![](imgs/developer-workflows.png)
+
+In addition to uncovering tools, this group should champion best practices for building developer experiences (DevEx) that enhance productivity and minimize cognitive load. This includes promoting tools that integrate seamlessly within existing workflows and empowering teams to design custom interfaces tailored to domain-specific needs.
+
+### Areas In-Scope
+
+A core objective of this group is to identify and promote existing tools, standards, and practices within the CNCF ecosystem. By showcasing how these elements work together, we aim to map how existing tools and standards can enhance and complement practices and tools already familiar to developers and architects.
```
Comment (TheFoxAtWork) [inline app-development-wg/charter/README.md:77]:
```suggestion
A core objective of this group is to identify and promote existing tools, specifications, and practices within the CNCF ecosystem. By showcasing how these elements work together, we aim to map how existing tools and standards can enhance and complement practices and tools already familiar to developers and architects.
```
CNCF is not a standards body, specifications are how we gain alignment, consistency, and exchange between tools and projects.

File: app-development-wg/charter/README.md [hunk 9]
```diff
@@ -0,0 +1,120 @@
...
+
+In addition to identifying best practices, this group should champion open-source tools that promote seamless integration between development tools and adopt standards. These tools empower teams to construct custom interfaces tailored to their specific domains, ultimately reducing cognitive load and boosting developer productivity.
```
Comment (TheFoxAtWork) [inline app-development-wg/charter/README.md:79]:
```suggestion
In addition to identifying best practices, this group should champion open source tools that promote seamless integration between development tools and adopt specifications. These tools empower teams to construct custom interfaces tailored to their specific domains, ultimately reducing cognitive load and boosting developer productivity.
```

File: app-development-wg/charter/README.md [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+- Provide early feedback and recommendations for projects and tools seeking CNCF adoption around cloud-native app development. 
+
+
+
+### Areas Out of Scope
+
+These are topics or initiatives that this WG will not target and work on. We might add new items or update this list if we find projects, standards, or initiatives this WG finds interesting to work on in future iterations. 
+- Create new standards.
+- Define how application development should be done.
+- Creation of a new CNCF open-source project.
```
Comment (TheFoxAtWork) [inline app-development-wg/charter/README.md:97]:
```suggestion
- Creation of a new CNCF open source project.
```

File: app-development-wg/charter/README.md (hunk 9 above)
Comment (TheFoxAtWork) [reply app-development-wg/charter/README.md:79]: this phrasin (adopt standards) doesnt quite make sense to me, is the intent  "integration between development tools and the specifications and practices necessary for application development to occur"?

File: app-development-wg/charter/README.md [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+
+In addition to identifying best practices, this group should champion open-source tools that promote seamless integration between development tools and adopt standards. These tools empower teams to construct custom interfaces tailored to their specific domains, ultimately reducing cognitive load and boosting developer productivity.
+
+The following list of goals will drive the working group agenda:
+
+- Identify and promote developer productivity tools within the cloud-native ecosystem. Raise awareness and facilitate the adoption of tools supporting and enhancing developer workflows in the cloud-native space.
+- Promote interoperability and standards across tools. Leverage shared APIs and SDKs across the industry to map the development ecosystem, document existing functionalities, and pinpoint areas for improvement.
+- Advocate for developer experience (DevEx) tools that streamline cloud-native development workflows.
+Assist with the organization of AppDeveloperCon, reviewing and proposing topics.
```
Comment (TheFoxAtWork) [inline app-development-wg/charter/README.md:86]:
```suggestion
Assist with the organization of Linux Foundation and CNCF events such as AppDeveloperCon, reviewing and proposing topics.
```
Incorporating the comment from the google doc version.

File: app-development-wg/charter/README.md [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+
+
+### Areas Out of Scope
+
+These are topics or initiatives that this WG will not target and work on. We might add new items or update this list if we find projects, standards, or initiatives this WG finds interesting to work on in future iterations. 
+- Create new standards.
+- Define how application development should be done.
+- Creation of a new CNCF open-source project.
+- Promote tools that are only useful for a specific stack/programming language.
+- Promote adopting security best practices throughout the development lifecycle, secure coding standards, and vulnerability scanning to mitigate risks and protect applications from potential threats.
```
Comment (TheFoxAtWork) [inline app-development-wg/charter/README.md:99]: I dont think this should be out of scope.  While TAG Security is largely responsible for this area, this working group should champion and consider secure coding practices (not standards), incorporation of vulnerability remediation, and secure design methodologies as part of their work. This is a supporting function, and will leverage much coordination and existing material from TAG Security but separating secure application development from application development further fractures both domains.  

I would suggest rephrasing this and moving it to in scope as:
```suggestion
- Advocate for secure application development practices as defined by TAG Security alongside any recommendations or material this group produces.
```

Comment (TheFoxAtWork) [reply app-development-wg/charter/README.md:99]: reading further, under "interface with other TAGs" such a change as proposed only serves to reinforce that collaboration.

File: app-development-wg/charter/README.md [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+- Promote sustainability and efficiency in application development and tooling to optimize resource utilization and minimize environmental impact.
```
Comment (TheFoxAtWork) [inline app-development-wg/charter/README.md:100]: Again, i don't think this should be out of scope.  We're facing a new application development reality that we (collectively as an industry) have forgotten due to overconsumption through resource proliferation.  It used to be that application developers were painfully aware of the very limited resources (such as available memory) to their applications and had to work around this as best they could (1996 RAM: 8mb, 2015 RAM 16gb).

Would recommend moving this up to in scope with the following suggested language:
```suggestion
- Advocate for sustainable and efficient application development practices as defined by TAG Environmental Sustainability alongside any recommendations or material this group produces.
```

Comment (TheFoxAtWork) [reply app-development-wg/charter/README.md:100]: reading further, under "interface with other TAGs" such a change as proposed only serves to reinforce that collaboration.

**Review by TheFoxAtWork — CHANGES_REQUESTED** [review]
Requested several changes for terminology consistency as well as scope/clarification.

File: app-development-wg/charter/README.md [hunk 14]
```diff
@@ -0,0 +1,120 @@
...
+
+- Local cloud-native runtime: a scaled-down version of the runtime is provided for developers to run locally in their workstations. 
+  - Using Kubernetes as the base for a cloud-native runtime, teams can run a scaled-down Kubernetes cluster locally with enough services to mimic production behavior for development tasks. Tools like MiniKube, KinD, and K3s are popular in this space. 
+- Ephemeral (remote) cloud-native runtime: a virtual scaled-down environment can be provisioned on-demand for a team to work on their development tasks. These environments are easy to create and easy to dispose of when they are no longer needed.
+  - If Kubernetes is being used, a platform team will be in charge of making the process simple and cost-efficient. 
+- Full remote cloud-native runtime: provide developers access to remote cloud-native runtime, where all the tools needed for their tasks are hosted remotely.
```
[the remark this answers is not in the document]
Comment (danieloh30) [reply app-development-wg/charter/README.md:23]: LGTM

File: app-development-wg/charter/README.md (hunk 8 above)
Comment (danieloh30) [reply app-development-wg/charter/README.md:77]: Just to be clear, CNCF doesn't suggest or represent any (industrial/official) standards, right? If so, any WGs can't propose or suggest standards (from the terminology standpoint), can they?

File: app-development-wg/charter/README.md (hunk 9 above)
Comment (danieloh30) [reply app-development-wg/charter/README.md:79]: Good catch! I rephrased the paragraph to be more clear! Let me know if that works for you @TheFoxAtWork 

In addition, this group should go beyond identifying best practices. They should actively advocate for open source tools that enable frictionless integration between development tools and adherence to established specifications. These tools empower teams to build custom interfaces specific to their domains, which ultimately reduces mental strain (cognitive load) and significantly boosts developer productivity.

File: app-development-wg/charter/README.md (hunk 8 above)
Comment (TheFoxAtWork) [reply app-development-wg/charter/README.md:77]: To my knowledge, @caniszczyk please correct me, CNCF does not represent in standards bodies. Regarding proposing or suggesting, some standards are for fees, some are industry specific (payment transactions for instance) so while we can enable adopters and projects to meet them by our understanding of those standards, proposing something like "cloud native projects should meet PCI/DSS" would be out of scope, but adopting a best practice like "secure software development" would be. 

Now, all this being said, if a project wants to provide a mechanism to support adopters in meeting a standard they can, and our groups are welcome to identify requirements of certain industry to help project reach adoption, referencing a standard to do so - but we don't create them.

If Chris has other opinions or perspective I'm keen to update my understanding.

File: app-development-wg/charter/README.md (hunk 9 above)
Comment (TheFoxAtWork) [reply app-development-wg/charter/README.md:79]: That reads much much better thank you!!

Comment (danieloh30) [reply app-development-wg/charter/README.md:79]: Cool! @salaboy can you update the PR with "In addition, this group should go beyond identifying best practices. They should actively advocate for open source tools that enable frictionless integration between development tools and adherence to established specifications. These tools empower teams to build custom interfaces specific to their domains, which ultimately reduces mental strain (cognitive load) and significantly boosts developer productivity."?

File: app-development-wg/charter/README.md (hunk 11 above)
Comment (danieloh30) [reply app-development-wg/charter/README.md:86]: LGTM!

**Review by lianmakesthings — APPROVED** [review]
Approved by the TOC in https://github.com/cncf/toc/issues/1311.
Congratulations to the team! Excited to see what will come out of this WG 🚀

**Pull request merged**